File: go/storage/mkvs/urkel/debug.go
package urkel

import (
	"context"
	"fmt"
	"io"
	"strings"

	"github.com/oasislabs/ekiden/go/storage/mkvs/urkel/node"
)

func (t *Tree) doDump(ctx context.Context, w io.Writer, ptr *node.Pointer, bitDepth node.Depth, path node.Key, depth node.Depth, right bool) {
	prefix := strings.Repeat(" ", int(depth)*2)
	nd, err := t.cache.derefNodePtr(ctx, node.ID{Path: path.AppendBit(bitDepth, right), BitDepth: bitDepth}, ptr, nil)
	if err != nil {
		panic(err)
	}

	switch n := nd.(type) {
	case nil:
		fmt.Fprint(w, prefix+"<nil>")
	case *node.InternalNode:
		fmt.Fprintf(w, prefix+"* [%v/%q(%d)/%s]: {\n", n.Clean, n.Label, n.LabelBitLength, n.Hash.String())
		// NB: depth+1 for LeafNode is purely for nicer indents. LeafNode should have the same depth as parent though.
		t.doDump(ctx, w, n.LeafNode, bitDepth+n.LabelBitLength, path.Merge(bitDepth, n.Label, n.LabelBitLength), depth+1, false)
		fmt.Fprintln(w, ",")
		t.doDump(ctx, w, n.Left, bitDepth+n.LabelBitLength, path.Merge(bitDepth, n.Label, n.LabelBitLength), depth+1, false)
		fmt.Fprintln(w, ",")
		t.doDump(ctx, w, n.Right, bitDepth+n.LabelBitLength, path.Merge(bitDepth, n.Label, n.LabelBitLength), depth+1, true)
		fmt.Fprintln(w, "")
		fmt.Fprint(w, prefix+"}")
	case *node.LeafNode:
		value, err := t.cache.derefValue(ctx, n.Value)
		if err != nil {
			value = []byte(fmt.Sprintf("<ERROR: %s>", err))
		}

		fmt.Fprintf(w, "%s- %s -> %s [%v/%s]", prefix, n.Key, value, n.Clean, n.Hash.String())
	default:
		fmt.Fprintf(w, prefix+"<UNKNOWN>")
	}
}

func (t *Tree) doStats(ctx context.Context, s *Stats, ptr *node.Pointer, bitDepth node.Depth, path node.Key, depth node.Depth, maxDepth node.Depth, right bool) node.Depth {
	if maxDepth > 0 && depth > maxDepth {
		return depth
	}
	if depth > s.MaxDepth {
		s.MaxDepth = depth
	}

	nd, err := t.cache.derefNodePtr(ctx, node.ID{Path: path.AppendBit(bitDepth, right), BitDepth: bitDepth}, ptr, nil)
	if err != nil {
		panic(err)
	}

	switch n := nd.(type) {
	case nil:
		s.DeadNodeCount++
	case *node.InternalNode:
		s.InternalNodeCount++

		leftDepth := t.doStats(ctx, s, n.Left, bitDepth+n.LabelBitLength, path.Merge(bitDepth, n.Label, n.LabelBitLength), depth+1, maxDepth, false)
		if leftDepth-depth > s.LeftSubtreeMaxDepths[depth] {
			s.LeftSubtreeMaxDepths[depth] = leftDepth - depth
		}

		rightDepth := t.doStats(ctx, s, n.Right, bitDepth+n.LabelBitLength, path.Merge(bitDepth, n.Label, n.LabelBitLength), depth+1, maxDepth, true)
		if rightDepth-depth > s.RightSubtreeMaxDepths[depth] {
			s.RightSubtreeMaxDepths[depth] = rightDepth - depth
		}

		if leftDepth > rightDepth {
			return leftDepth
		}
		return rightDepth
	case *node.LeafNode:
		value, err := t.cache.derefValue(ctx, n.Value)
		if err != nil {
			panic(err)
		}

		s.LeafNodeCount++
		s.LeafValueSize += uint64(len(value))
	}

	return depth
}

func (t *Tree) doDumpLocal(ctx context.Context, w io.Writer, ptr *node.Pointer, depth node.Depth, maxDepth node.Depth) {
	prefix := strings.Repeat(" ", int(depth)*2)
	if ptr == nil {
		fmt.Fprint(w, prefix+"<nil>")
		return
	}

	if maxDepth > 0 && depth > maxDepth {
		fmt.Fprint(w, prefix+"<...>")
		return
	}

	nd := ptr.Node

	switch n := nd.(type) {
	case nil:
		fmt.Fprintf(w, prefix+"<nil> [%v/%s]", ptr.Clean, ptr.Hash)
	case *node.InternalNode:
		fmt.Fprintf(w, prefix+"* [%v/%q(%d)/%s]: {\n", n.Clean, n.Label, n.LabelBitLength, n.Hash)
		// NB: depth+1 for LeafNode is purely for nicer indents. LeafNode should have the same depth as parent though.
		t.doDumpLocal(ctx, w, n.LeafNode, depth+1, maxDepth)
		fmt.Fprintln(w, ",")
		t.doDumpLocal(ctx, w, n.Left, depth+1, maxDepth)
		fmt.Fprintln(w, ",")
		t.doDumpLocal(ctx, w, n.Right, depth+1, maxDepth)
		fmt.Fprintln(w, "")
		fmt.Fprint(w, prefix+"}")
	case *node.LeafNode:
		value := n.Value

		fmt.Fprintf(w, "%s- %s -> %v [%v/%s]", prefix, n.Key, value, n.Clean, n.Hash)
	default:
		fmt.Fprintf(w, prefix+"<UNKNOWN>")
	}
}

File: runtime/src/storage/mkvs/urkel/tree/debug.rs
use std::{cmp::max, collections::BTreeMap, iter::repeat, sync::Arc};

use io_context::Context;

use crate::storage::mkvs::urkel::{cache::*, tree::*};

impl UrkelTree {
    /// Traverse the tree and return some statistics about it.
    pub fn stats(&mut self, ctx: Context, max_depth: Depth) -> UrkelStats {
        let ctx = ctx.freeze();
        let mut stats = UrkelStats {
            left_subtree_max_depths: BTreeMap::new(),
            right_subtree_max_depths: BTreeMap::new(),
            cache: self.cache.borrow_mut().stats(),
            ..Default::default()
        };
        let pending_root = self.cache.borrow().get_pending_root();
        self._stats(
            &ctx,
            &mut stats,
            pending_root,
            0,
            Key::new(),
            0,
            max_depth,
            false,
        );
        stats
    }

    fn _stats(
        &mut self,
        ctx: &Arc<Context>,
        stats: &mut UrkelStats,
        ptr: NodePtrRef,
        bit_depth: Depth,
        path: Key,
        depth: Depth,
        max_depth: Depth,
        right: bool,
    ) -> Depth {
        if max_depth > 0 && depth > max_depth {
            return depth;
        }

        if depth > stats.max_depth {
            stats.max_depth = depth;
        }

        let node_ref = self
            .cache
            .borrow_mut()
            .deref_node_ptr(
                ctx,
                NodeID {
                    path: &path.append_bit(bit_depth, right),
                    bit_depth: bit_depth,
                },
                ptr,
                None,
            )
            .ok()
            .unwrap_or(None);

        match classify_noderef!(?node_ref) {
            NodeKind::None => {
                stats.dead_node_count += 1;
            }
            NodeKind::Internal => {
                let node_ref = node_ref.unwrap();
                stats.internal_node_count += 1;

                let left_depth = self._stats(
                    ctx,
                    stats,
                    noderef_as!(node_ref, Internal).left.clone(),
                    bit_depth + noderef_as!(node_ref, Internal).label_bit_length,
                    path.merge(
                        bit_depth,
                        &noderef_as!(node_ref, Internal).label,
                        noderef_as!(node_ref, Internal).label_bit_length,
                    ),
                    depth + 1,
                    max_depth,
                    false,
                );
                if left_depth - depth > *stats.left_subtree_max_depths.get(&depth).unwrap_or(&0) {
                    stats
                        .left_subtree_max_depths
                        .insert(depth, left_depth - depth);
                }

                let right_depth = self._stats(
                    ctx,
                    stats,
                    noderef_as!(node_ref, Internal).right.clone(),
                    bit_depth + noderef_as!(node_ref, Internal).label_bit_length,
                    path.merge(
                        bit_depth,
                        &noderef_as!(node_ref, Internal).label,
                        noderef_as!(node_ref, Internal).label_bit_length,
                    ),
                    depth + 1,
                    max_depth,
                    true,
                );
                if right_depth - depth > *stats.right_subtree_max_depths.get(&depth).unwrap_or(&0) {
                    stats
                        .right_subtree_max_depths
                        .insert(depth, right_depth - depth);
                }

                return max(left_depth, right_depth);
            }
            NodeKind::Leaf => {
                let node_ref = node_ref.unwrap();
                let value = self
                    .cache
                    .borrow_mut()
                    .deref_value_ptr(ctx, noderef_as!(node_ref, Leaf).value.clone());
                let value = match value {
                    Err(err) => panic!("{}", err),
                    Ok(value) => value,
                };

                stats.leaf_node_count += 1;
                if let Some(ref value) = value {
                    stats.leaf_value_size += value.len();
                }
            }
        };

        depth
    }

    /// Get an object that can be formatted using `{:#?}`.
    pub fn get_dumpable(&mut self) -> NodePtrRef {
        self.cache.borrow().get_pending_root().clone()
    }

    /// Dump the tree into the given writer.
    pub fn dump(&mut self, ctx: Context, w: &mut impl ::std::io::Write) -> ::std::io::Result<()> {
        let ctx = ctx.freeze();
        let pending_root = self.cache.borrow().get_pending_root();
        self._dump(&ctx, w, pending_root, 0, Key::new(), 0, false)?;
        writeln!(w, "")
    }

    fn _dump(
        &mut self,
        ctx: &Arc<Context>,
        w: &mut impl ::std::io::Write,
        ptr: NodePtrRef,
        bit_depth: Depth,
        path: Key,
        depth: Depth,
        right: bool,
    ) -> ::std::io::Result<()> {
        let prefix = repeat(" ").take(depth as usize * 2).collect::<String>();

        let node = self
            .cache
            .borrow_mut()
            .deref_node_ptr(
                ctx,
                NodeID {
                    path: &path.append_bit(bit_depth, right),
                    bit_depth: bit_depth,
                },
                ptr,
                None,
            )
            .ok()
            .unwrap_or(None);

        match classify_noderef!(?node) {
            NodeKind::None => write!(w, "{}<nil>", prefix),
            NodeKind::Internal => {
                let some_node_ref = match node {
                    Some(ref node) => node,
                    _ => unreachable!(),
                };
                write!(
                    w,
                    "{}+ [{}/{:?}({})/{:?}]: {{\n",
                    prefix,
                    noderef_as!(some_node_ref, Internal).clean,
                    noderef_as!(some_node_ref, Internal).label,
                    noderef_as!(some_node_ref, Internal).label_bit_length,
                    noderef_as!(some_node_ref, Internal).hash
                )?;

                self._dump(
                    ctx,
                    w,
                    noderef_as!(some_node_ref, Internal).leaf_node.clone(),
                    bit_depth + noderef_as!(some_node_ref, Internal).label_bit_length,
                    path.merge(
                        bit_depth,
                        &noderef_as!(some_node_ref, Internal).label,
                        noderef_as!(some_node_ref, Internal).label_bit_length,
                    ),
                    // NB: depth+1 for LeafNode is purely for nicer indents. LeafNode should have the same depth as parent though.
                    depth + 1,
                    false,
                )?;
                writeln!(w, ",")?;
                self._dump(
                    ctx,
                    w,
                    noderef_as!(some_node_ref, Internal).left.clone(),
                    bit_depth + noderef_as!(some_node_ref, Internal).label_bit_length,
                    path.merge(
                        bit_depth,
                        &noderef_as!(some_node_ref, Internal).label,
                        noderef_as!(some_node_ref, Internal).label_bit_length,
                    ),
                    depth + 1,
                    false,
                )?;
                writeln!(w, ",")?;
                self._dump(
                    ctx,
                    w,
                    noderef_as!(some_node_ref, Internal).right.clone(),
                    bit_depth + noderef_as!(some_node_ref, Internal).label_bit_length,
                    path.merge(
                        bit_depth,
                        &noderef_as!(some_node_ref, Internal).label,
                        noderef_as!(some_node_ref, Internal).label_bit_length,
                    ),
                    depth + 1,
                    true,
                )?;
                writeln!(w, "")?;
                write!(w, "{}}}", prefix)
            }
            NodeKind::Leaf => {
                let some_node_ref = match node {
                    Some(ref node) => node,
                    _ => unreachable!(),
                };
                let value = self
                    .cache
                    .borrow_mut()
                    .deref_value_ptr(ctx, noderef_as!(some_node_ref, Leaf).value.clone());
                match value {
                    Err(err) => write!(w, "<ERROR: {}>", err.as_fail()),
                    Ok(value) => write!(
                        w,
                        "{}- {:?} -> {:?} [{}/{:?}]",
                        prefix,
                        noderef_as!(some_node_ref, Leaf).key,
                        value,
                        noderef_as!(some_node_ref, Leaf).clean,
                        noderef_as!(some_node_ref, Leaf).hash
                    ),
                }
            }
        }
    }
}
